File: examples/java/com/intel/daal/examples/gbt/GbtRegTraversedModelBuilder.java
/* file: GbtRegTraversedModelBuilder.java */
/*******************************************************************************
* Copyright 2014-2019 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
 //  Content:
 //     Java example of gradient boosted trees regression model
 //     building from traversed gradient boosted trees regression model.
 //
 //     The program trains the gradient boosted trees regression model, gets
 //     pre-computed values from nodes of each tree using traversal and build
 //     model of the gradient boosted trees regression via Model Builder and
 //     computes regression for the test data.
 ////////////////////////////////////////////////////////////////////////////////
 */

/**
 * <a name="DAAL-EXAMPLE-JAVA-GBTREGTRAVERSEDMODELBUILDER">
 * @example GbtRegTraversedModelBuilder.java
 */

package com.intel.daal.examples.gbt;

import com.intel.daal.algorithms.tree_utils.regression.TreeNodeVisitor;
import com.intel.daal.algorithms.tree_utils.regression.LeafNodeDescriptor;
import com.intel.daal.algorithms.tree_utils.SplitNodeDescriptor;
import com.intel.daal.algorithms.gbt.regression.training.InputId;
import com.intel.daal.algorithms.gbt.regression.training.TrainingResultId;
import com.intel.daal.algorithms.gbt.regression.prediction.ModelInputId;
import com.intel.daal.algorithms.gbt.regression.prediction.NumericTableInputId;
import com.intel.daal.algorithms.gbt.regression.prediction.PredictionResult;
import com.intel.daal.algorithms.gbt.regression.prediction.PredictionResultId;
import com.intel.daal.algorithms.gbt.regression.Model;
import com.intel.daal.algorithms.gbt.regression.ModelBuilder;
import com.intel.daal.algorithms.gbt.regression.prediction.*;
import com.intel.daal.algorithms.gbt.regression.training.*;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.data_management.data.HomogenNumericTable;
import com.intel.daal.data_management.data.MergedNumericTable;
import com.intel.daal.data_management.data_source.DataSource;
import com.intel.daal.data_management.data_source.FileDataSource;
import com.intel.daal.examples.utils.Service;
import com.intel.daal.services.DaalContext;
import com.intel.daal.data_management.data.*;
import java.util.LinkedList;
import java.util.Queue;


/* Node structure for representing nodes in trees after traversing DAAL model */
class NodeReg {
    public NodeReg left;
    public NodeReg right;
    public double response;
    public long featureIndex;
    public double featureValue;

    NodeReg(double rs, long fi, double fv) {
        left = null;
        right = null;
        response = rs;
        featureIndex = fi;
        featureValue = fv;
    }

    NodeReg() {
        left = null;
        right = null;
        response = 0;
        featureIndex = 0;
        featureValue = 0;
    }
}

/* Tree structure for representing tree after traversing DAAL model */
class TreeReg {
    public NodeReg root;
    public long nNodes;
}

/* Example of structure to remember relationship between nodes */
class ParentPlaceReg {
    public long parentId;
    public long place;

    ParentPlaceReg(long _parent, long _place) {
        parentId = _parent;
        place = _place;
    }

    ParentPlaceReg() {
        parentId = 0;
        place = 0;
    }
}

/* Visitor class implementing TreeNodeVisitor interface, prints out tree nodes of the model when it is called back by model traversal method */
class BFSNodeVisitorReg extends TreeNodeVisitor {

    public TreeReg [] roots;
    int treeId;
    Queue<NodeReg> parentNodes;

    BFSNodeVisitorReg(int nTrees) {
        roots = new TreeReg[nTrees];
        for(int i = 0; i < nTrees; i++) {
            roots[i] = new TreeReg();
            roots[i].root = new NodeReg();
        }
        treeId = 0;
        parentNodes = new LinkedList<NodeReg>();
    }

    @Override
    public boolean onLeafNode(LeafNodeDescriptor desc) {
        if(desc.level == 0) {
            NodeReg root = roots[treeId].root;
            roots[treeId].nNodes = 1;
            root.left = null;
            root.right = null;
            root.response = desc.response;
            root.featureIndex = 0;
            root.featureValue = 0;
            treeId++;
        }
        else {
            roots[treeId - 1].nNodes++;
            NodeReg node = new NodeReg(desc.response, 0, 0);

            NodeReg parent = parentNodes.peek();
            if(parent.left == null) {
                parent.left = node;
            }
            else {
                parent.right = node;
                parentNodes.poll();
            }
        }
        return true;
    }

    @Override
    public boolean onSplitNode(SplitNodeDescriptor desc) {
        if(desc.level == 0) {
            NodeReg root = roots[treeId].root;
            roots[treeId].nNodes = 1;
            root.left = null;
            root.right = null;
            root.response = 0;
            root.featureIndex = desc.featureIndex;
            root.featureValue = desc.featureValue;
            parentNodes.add(root);
            treeId++;
        }
        else {
            roots[treeId - 1].nNodes++;
            NodeReg node = new NodeReg(0,desc.featureIndex,desc.featureValue);

            NodeReg parent = parentNodes.peek();
            if(parent.left == null) {
                parent.left = node;
            }
            else {
                parent.right = node;
                parentNodes.poll();
            }
            parentNodes.add(node);
        }
        return true;
    }

}

class GbtRegTraversedModelBuilder {
    /* Input data set parameters */
    private static final String trainDatasetFileName = "../data/batch/df_regression_train.csv";
    private static final String testDatasetFileName = "../data/batch/df_regression_test.csv";
    private static final long categoricalFeaturesIndices [] =  { 3 };

    /* Number of features in training and testing data sets */
    private static final int nFeatures     = 13;
    /* Gradient boosted trees training parameters */
    private static int nTrees = 0;
    private static final int maxIterations = 40;

    private static DaalContext context = new DaalContext();

    public static void main(String[] args) throws java.io.FileNotFoundException, java.io.IOException {

        /* Train DAAL Gradient Boosted Trees Regression model */
        TrainingResult trainingResult = trainModel();
        System.out.println("Predict on trained model");
        Model trainedModel = trainingResult.get(TrainingResultId.model);
        if(trainedModel != null) {
            nTrees = (int)trainedModel.getNumberOfTrees();
        }
        double trainedAccurcy = testModel(trainedModel);

        /* Traverse the trained model to get Tree representation */
        BFSNodeVisitorReg visitor = new BFSNodeVisitorReg(nTrees);
        TreeReg [] trees = traverseModel(trainedModel, visitor);
        /* Build the model by ModelBuilder from Tree */
        Model builtModel = buildModel(trees);
        System.out.println("Predict on built model from input user Tree ");
        double buildModelAccurcy = testModel(builtModel);

        if(trainedAccurcy == buildModelAccurcy) {
            System.out.println("Model was built successfully");
        }
        else {
            System.out.println("Model was built not correctly");
        }

        context.dispose();
    }

    public static Model buildModel(TreeReg [] trees) {

        /* Create a model builder */
        ModelBuilder modelBuilder = new ModelBuilder(context, nFeatures, nTrees);

        for(int i = 0; i < nTrees; i++) {
            final long nNodes = trees[i].nNodes;

            /* Allocate the memory for certain tree */
            modelBuilder.createTree(nNodes);
            boolean isRoot = true;

            /* Recursive DFS traversing of certain tree with building model */
            buildTree(i, trees[i].root, isRoot, modelBuilder, new ParentPlaceReg(0,0));
        }

        return modelBuilder.getModel();
    }

    private static boolean buildTree(long treeId, NodeReg node, boolean isRoot, ModelBuilder builder, ParentPlaceReg p) {

        if(node.left != null && node.right != null) {
            if(isRoot) {
                long parent = builder.addSplitNode(treeId, ModelBuilder.noParent, 0, node.featureIndex, node.featureValue);

                isRoot = false;
                buildTree(treeId, node.left, isRoot, builder, new ParentPlaceReg(parent, 0));
                buildTree(treeId, node.right, isRoot, builder, new ParentPlaceReg(parent, 1));
            }
            else {
                long parent = builder.addSplitNode(treeId, p.parentId, p.place, node.featureIndex, node.featureValue);

                buildTree(treeId, node.left, isRoot, builder, new ParentPlaceReg(parent, 0));
                buildTree(treeId, node.right, isRoot, builder, new ParentPlaceReg(parent, 1));
            }
        }
        else {
            if(isRoot) {
                builder.addLeafNode(treeId, ModelBuilder.noParent, 0, node.response);
                isRoot = false;
            }
            else {
                builder.addLeafNode(treeId, p.parentId, p.place, node.response);
            }
            return true;
        }

        return true;
    }

    private static TrainingResult trainModel() {
        /* Retrieve the data from the input data sets */
        FileDataSource trainDataSource = new FileDataSource(context, trainDatasetFileName,
                DataSource.DictionaryCreationFlag.DoDictionaryFromContext,
                DataSource.NumericTableAllocationFlag.NotAllocateNumericTable);

        /* Create Numeric Tables for training data and labels */
        NumericTable trainData = new HomogenNumericTable(context, Float.class, nFeatures, 0, NumericTable.AllocationFlag.NotAllocate);
        NumericTable trainGroundTruth = new HomogenNumericTable(context, Float.class, 1, 0, NumericTable.AllocationFlag.NotAllocate);
        MergedNumericTable mergedData = new MergedNumericTable(context);
        mergedData.addNumericTable(trainData);
        mergedData.addNumericTable(trainGroundTruth);

        /* Retrieve the data from an input file */
        trainDataSource.loadDataBlock(mergedData);

        /* Set feature as categorical */
        trainData.getDictionary().setFeature(Float.class,3,DataFeatureUtils.FeatureType.DAAL_CATEGORICAL);

        /* Create algorithm objects to train the gradient boosted trees regression model */
        TrainingBatch algorithm = new TrainingBatch(context, Float.class, TrainingMethod.defaultDense);

        /* Pass a training data set and dependent values to the algorithm */
        algorithm.input.set(InputId.data, trainData);
        algorithm.input.set(InputId.dependentVariable, trainGroundTruth);
        algorithm.parameter.setMaxIterations(maxIterations);

        /* Train the decision forest regression model */
        return algorithm.compute();
    }

    private static double testModel(Model model) {

        /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
        FileDataSource testDataSource = new FileDataSource(context, testDatasetFileName,
                DataSource.DictionaryCreationFlag.DoDictionaryFromContext,
                DataSource.NumericTableAllocationFlag.NotAllocateNumericTable);

        /* Create Numeric Tables for testing data and labels */
        NumericTable testData = new HomogenNumericTable(context, Float.class, nFeatures, 0, NumericTable.AllocationFlag.NotAllocate);
        NumericTable testGroundTruth = new HomogenNumericTable(context, Float.class, 1, 0, NumericTable.AllocationFlag.NotAllocate);
        MergedNumericTable mergedData = new MergedNumericTable(context);
        mergedData.addNumericTable(testData);
        mergedData.addNumericTable(testGroundTruth);

        /* Retrieve the data from an input file */
        testDataSource.loadDataBlock(mergedData);
        /* Set feature as categorical */
        testData.getDictionary().setFeature(Float.class, 3, DataFeatureUtils.FeatureType.DAAL_CATEGORICAL);

        /* Create algorithm objects for gradient boosted trees regression prediction */
        PredictionBatch algorithm = new PredictionBatch(context, Float.class, PredictionMethod.defaultDense);

        /* Pass a testing data set and the trained model to the algorithm */
        algorithm.input.set(NumericTableInputId.data, testData);
        algorithm.input.set(ModelInputId.model, model);

        /* Compute prediction results */
        PredictionResult predictionResult = algorithm.compute();
        NumericTable prediction = predictionResult.get(PredictionResultId.prediction);
        Service.printNumericTable("Gradient boosted trees prediction results (first 10 rows):", prediction, 10);
        Service.printNumericTable("Ground truth (first 10 rows):", testGroundTruth, 10);

        long nRows = 0;
        if(prediction != null) {
            nRows = prediction.getNumberOfRows();
        }

        double error = 0;
        for(long i = 0; i < nRows; i++) {
                error += (prediction.getFloatValue(0,i) - testGroundTruth.getFloatValue(0,i));
        }

        System.out.print("Error: ");
        System.out.format("%.3f%n", error);

        return error;
    }

    private static TreeReg [] traverseModel(Model model, BFSNodeVisitorReg visitor) {
        final int nTrees = (int)model.getNumberOfTrees();

        for(int i = 0; i < nTrees; ++i) {
            model.traverseBFS(i, visitor);
        }
        return visitor.roots;
    }

}

File: examples/daal/cpp/source/gradient_boosted_trees/gbt_reg_traversed_model_builder.cpp
/* file: gbt_reg_traversed_model_builder.cpp */
/*******************************************************************************
* Copyright 2014 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of gradient boosted trees regression model
!    building from traversed gradient boosted trees regression model.
!
!    The program trains the gradient boosted trees regression model, gets
!    pre-computed values from nodes of each tree using traversal and build
!    model of the gradient boosted trees regression via Model Builder and
!    computes regression for the test data.
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-GBT_REG_TRAVERSED_MODEL_BUILDER"></a>
 * \example gbt_reg_traversed_model_builder.cpp
 */

#include "daal.h"
#include "service.h"

using namespace daal;
using namespace daal::data_management;
using namespace daal::algorithms::gbt::regression;

/* Input data set parameters */
const std::string trainDatasetFileName = "../data/batch/df_regression_train.csv";
const std::string testDatasetFileName = "../data/batch/df_regression_test.csv";
const size_t categoricalFeaturesIndices[] = { 3 };
const size_t nFeatures = 13; /* Number of features in training and testing data sets */

/* Gradient boosted trees training parameters */
const size_t maxIterations = 40;
size_t nTrees = 0;

/** Node structure for representing nodes in trees after traversing DAAL model */
struct Node {
    Node *left;
    Node *right;
    double response;
    size_t featureIndex;
    double featureValue;
    Node(double rs, size_t fi, double fv)
            : left(NULL),
              right(NULL),
              response(rs),
              featureIndex(fi),
              featureValue(fv) {}
    Node() : left(NULL), right(NULL), response(0), featureIndex(0), featureValue(0) {}
};

/** Tree structure for representing tree after traversing DAAL model */
struct Tree {
    Node *root;
    size_t nNodes;
    ~Tree() {
        if (root) {
            delete root;
        }
    }
};

/** Example of structure to remember relationship between nodes */
struct ParentPlace {
    size_t parentId;
    size_t place;
    ParentPlace(size_t _parent, size_t _place) : parentId(_parent), place(_place) {}
    ParentPlace() : parentId(0), place(0) {}
};

/** Visitor class implementing TreeNodeVisitor interface, prints out tree nodes of the model when it is called back by model traversal method */
class BFSNodeVisitor : public daal::algorithms::tree_utils::regression::TreeNodeVisitor {
public:
    Tree *roots;
    size_t treeId;
    std::queue<Node *> parentNodes;
    virtual bool onLeafNode(
        const daal::algorithms::tree_utils::regression::LeafNodeDescriptor &desc) {
        if (desc.level == 0) {
            Node *root = roots[treeId].root;
            (*(roots + treeId)).nNodes = 1;
            root->left = NULL;
            root->right = NULL;
            root->response = desc.response;
            root->featureIndex = 0;
            root->featureValue = 0;
            treeId++;
        }
        else {
            roots[treeId - 1].nNodes++;
            Node *node = new Node(desc.response, 0, 0);

            Node *parent = parentNodes.front();
            if (parent->left == NULL) {
                parent->left = node;
            }
            else {
                parent->right = node;
                parentNodes.pop();
            }
        }
        return true;
    }

    virtual bool onSplitNode(
        const daal::algorithms::tree_utils::regression::SplitNodeDescriptor &desc) {
        if (desc.level == 0) {
            Node *root = roots[treeId].root;
            (*(roots + treeId)).nNodes = 1;
            root->left = NULL;
            root->right = NULL;
            root->response = 0;
            root->featureIndex = desc.featureIndex;
            root->featureValue = desc.featureValue;
            parentNodes.push(root);
            treeId++;
        }
        else {
            roots[treeId - 1].nNodes++;
            Node *node = new Node(0, desc.featureIndex, desc.featureValue);

            Node *parent = parentNodes.front();
            if (parent->left == NULL) {
                parent->left = node;
            }
            else {
                parent->right = node;
                parentNodes.pop();
            }
            parentNodes.push(node);
        }
        return true;
    }

    BFSNodeVisitor(size_t nTrees) : parentNodes() {
        roots = new Tree[nTrees];
        for (size_t i = 0; i < nTrees; i++) {
            roots[i].root = new Node;
        }
        treeId = 0;
    }
    ~BFSNodeVisitor() {
        if (roots)
            delete[] roots;
    }
};

training::ResultPtr trainModel();
double testModel(ModelPtr modelPtr);
void loadData(const std::string &fileName, NumericTablePtr &pData, NumericTablePtr &pDependentVar);
ModelPtr buildModel(Tree *trees);
Tree *traverseModel(ModelPtr m, BFSNodeVisitor &visitor);

bool buildTree(size_t treeId,
               Node *node,
               bool &isRoot,
               ModelBuilder &builder,
               const ParentPlace &parentPlace);

int main(int argc, char *argv[]) {
    checkArguments(argc, argv, 1, &trainDatasetFileName);

    /* Train DAAL Gradient Boosted Trees Regression model */
    training::ResultPtr trainingResult = trainModel();
    std::cout << "Predict on trained model" << std::endl;
    ModelPtr trainedModel = trainingResult->get(training::model);
    if (trainedModel.get())
        nTrees = trainedModel->numberOfTrees();
    double trainedAccurcy = testModel(trainedModel);

    /* Traverse the trained model to get Tree representation */
    BFSNodeVisitor visitor(nTrees);
    Tree *trees = traverseModel(trainedModel, visitor);
    /* Build the model by ModelBuilder from Tree */
    daal::algorithms::gbt::regression::ModelPtr builtModel = buildModel(trees);
    std::cout << "Predict on built model from input user Tree " << std::endl;
    double buildModelAccurcy = testModel(builtModel);
    const char *result = (trainedAccurcy == buildModelAccurcy) ? "successfully" : "not correctly";
    std::cout << "Model was built " << result << std::endl;

    return (trainedAccurcy == buildModelAccurcy) ? 0 : 1;
}

daal::algorithms::gbt::regression::ModelPtr buildModel(Tree *trees) {
    /* create a model builder */
    ModelBuilder builder(nFeatures, nTrees);
    for (size_t i = 0; i < nTrees; i++) {
        const size_t nNodes = trees[i].nNodes;
        /* allocate the memory for certain tree */
        builder.createTree(nNodes);
        bool isRoot = true;
        /* recursive DFS traversing of certain tree with building model */
        buildTree(i, trees[i].root, isRoot, builder, ParentPlace(0, 0));
    }

    return builder.getModel();
}

bool buildTree(size_t treeId,
               Node *node,
               bool &isRoot,
               ModelBuilder &builder,
               const ParentPlace &parentPlace) {

    const int defaultLeft = 0;
    const double cover = 0.0;
    if (node->left != NULL && node->right != NULL) {
        if (isRoot) {
            ModelBuilder::NodeId parent = builder.addSplitNode(treeId,
                                                               ModelBuilder::noParent,
                                                               0,
                                                               node->featureIndex,
                                                               node->featureValue,
                                                               defaultLeft,
                                                               cover);

            isRoot = false;
            buildTree(treeId, node->left, isRoot, builder, ParentPlace(parent, 0));
            buildTree(treeId, node->right, isRoot, builder, ParentPlace(parent, 1));
        }
        else {
            ModelBuilder::NodeId parent = builder.addSplitNode(treeId,
                                                               parentPlace.parentId,
                                                               parentPlace.place,
                                                               node->featureIndex,
                                                               node->featureValue,
                                                               defaultLeft,
                                                               cover);

            buildTree(treeId, node->left, isRoot, builder, ParentPlace(parent, 0));
            buildTree(treeId, node->right, isRoot, builder, ParentPlace(parent, 1));
        }
    }
    else {
        if (isRoot) {
            builder.addLeafNode(treeId, ModelBuilder::noParent, 0, node->response, cover);
            isRoot = false;
        }
        else {
            builder.addLeafNode(treeId, parentPlace.parentId, parentPlace.place, node->response, cover);
        }
    }

    return true;
}

double testModel(daal::algorithms::gbt::regression::ModelPtr modelPtr) {
    /* Create Numeric Tables for testing data and ground truth values */
    NumericTablePtr testData;
    NumericTablePtr testGroundTruth;

    loadData(testDatasetFileName, testData, testGroundTruth);

    /* Create an algorithm object to predict values of decision forest classification */
    prediction::Batch<> algorithm;

    /* Pass a testing data set and the trained model to the algorithm */
    algorithm.input.set(prediction::data, testData);
    algorithm.input.set(prediction::model, modelPtr);

    /* Predict values of decision forest classification */
    algorithm.compute();

    /* Retrieve the algorithm results */
    NumericTablePtr prediction = algorithm.getResult()->get(prediction::prediction);
    printNumericTable(prediction, "Gradient boosted trees prediction results (first 10 rows):", 10);
    printNumericTable(testGroundTruth, "Ground truth (first 10 rows):", 10);
    size_t nRows = 0;
    if (prediction.get())
        nRows = prediction->getNumberOfRows();
    double error = 0;
    for (size_t i = 0; i < nRows; i++) {
        error += prediction->getValue<float>(0, i) - testGroundTruth->getValue<float>(0, i);
    }

    std::cout << "Error: " << error << std::endl;
    return error;
}

training::ResultPtr trainModel() {
    /* Create Numeric Tables for training data and dependent variables */
    NumericTablePtr trainData;
    NumericTablePtr trainDependentVariable;

    loadData(trainDatasetFileName, trainData, trainDependentVariable);

    /* Create an algorithm object to train the decision forest classification model */
    training::Batch<> algorithm;

    /* Pass a training data set and dependent values to the algorithm */
    algorithm.input.set(training::data, trainData);
    algorithm.input.set(training::dependentVariable, trainDependentVariable);

    algorithm.parameter().maxIterations = maxIterations;

    /* Build the decision forest classification model */
    algorithm.compute();

    /* Retrieve the algorithm results */
    return algorithm.getResult();
}

void loadData(const std::string &fileName, NumericTablePtr &pData, NumericTablePtr &pDependentVar) {
    /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> trainDataSource(fileName,
                                                      DataSource::notAllocateNumericTable,
                                                      DataSource::doDictionaryFromContext);

    /* Create Numeric Tables for training data and dependent variables */
    pData.reset(new HomogenNumericTable<double>(nFeatures, 0, NumericTable::notAllocate));
    pDependentVar.reset(new HomogenNumericTable<double>(1, 0, NumericTable::notAllocate));
    NumericTablePtr mergedData(new MergedNumericTable(pData, pDependentVar));

    /* Retrieve the data from input file */
    trainDataSource.loadDataBlock(mergedData.get());

    NumericTableDictionaryPtr pDictionary = pData->getDictionarySharedPtr();
    for (size_t i = 0,
                n = sizeof(categoricalFeaturesIndices) / sizeof(categoricalFeaturesIndices[0]);
         i < n;
         ++i)
        (*pDictionary)[categoricalFeaturesIndices[i]].featureType =
            data_feature_utils::DAAL_CATEGORICAL;
}

Tree *traverseModel(const daal::algorithms::gbt::regression::ModelPtr m, BFSNodeVisitor &visitor) {
    const size_t nTrees = m->numberOfTrees();

    for (size_t i = 0; i < nTrees; ++i) {
        m->traverseBFS(i, visitor);
    }
    return visitor.roots;
}
